Dockerfile

```diff
@@ -10,31 +10,46 @@ RUN apt-get update \
         libssl-dev \
         git \
     && apt-get clean \
-    && rm -rf /var/lib/apt/lists/*
+    && rm -rf /var/lib/apt/lists/* \
+    && python -m venv /opt/venv
 
-COPY requirements.txt /opt/CTFd/
-
-RUN pip install -r requirements.txt --no-cache-dir
+ENV PATH="/opt/venv/bin:$PATH"
 
 COPY . /opt/CTFd
 
-# hadolint ignore=SC2086
-RUN for d in CTFd/plugins/*; do \
+RUN pip install --no-cache-dir -r requirements.txt \
+    && for d in CTFd/plugins/*; do \
         if [ -f "$d/requirements.txt" ]; then \
-            pip install -r $d/requirements.txt --no-cache-dir; \
+            pip install --no-cache-dir -r "$d/requirements.txt";\
         fi; \
     done;
 
-# hadolint ignore=DL3059
-RUN adduser \
-    --disabled-login \
-    -u 1001 \
-    --gecos "" \
+
+FROM python:3.9-slim-buster as release
+WORKDIR /opt/CTFd
+
+# hadolint ignore=DL3008
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+        libffi6 \
+        libssl1.1 \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY --chown=1001:1001 . /opt/CTFd
+
+RUN useradd \
```

Comment:
What is the advantage of switching to useradd?

Reply:
adduser is a wrapper for useradd binary.
First, useradd, by default, not generate home directory. It's not needed in this application.
Second, adduser not have option to remove log-init files. useradd have option to remove lastlog and faillog files, --no-log-init.
Third, using Dive application size is reduced for this change is 440k to 5.6k.

Reply:
Sounds good to me